semantic-analysis/src/Analysis/Syntax/Python.hs

```diff
@@ -1,82 +1,122 @@
-{-# LANGUAGE DeriveTraversable #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE LambdaCase #-}
 -- | This belongs in @semantic-python@ instead of @semantic-analysis@, but for the sake of expedience…
 module Analysis.Syntax.Python
 ( -- * Syntax
-  Term
-, Python(..)
+  Term(..)
+, subterms
   -- * Abstract interpretation
 , eval0
 , eval
+  -- * Parsing
+, parse
 ) where
 
 import           Analysis.Effect.Domain hiding ((:>>>))
 import qualified Analysis.Effect.Statement as S
 import           Analysis.Name
 import           Analysis.Reference
-import qualified Analysis.Syntax as T
 import           Analysis.VM
 import           Control.Effect.Labelled
 import           Control.Effect.Reader
+import           Control.Monad (foldM)
+import           Data.Foldable (for_)
 import           Data.Function (fix)
-import           Data.List.NonEmpty (NonEmpty)
-import           Data.Text (Text)
-import           Source.Span (Span)
+import           Data.List.NonEmpty (nonEmpty)
+import           Data.Maybe (mapMaybe)
+import qualified Data.Set as Set
+import           Data.Text (pack)
+import qualified Language.Python.Common as Py
+import           Language.Python.Version3.Parser
+import           Source.Span (Pos (..), Span (..), point)
+import           System.FilePath (takeBaseName)
 
 -- Syntax
 
-type Term = T.Term Python Name
+data Term
+  = Module (Py.Module Py.SrcSpan)
+  | Statement (Py.Statement Py.SrcSpan)
+  | Expr (Py.Expr Py.SrcSpan)
+  deriving (Eq, Ord, Show)
 
-data Python t
-  = Noop
-  | Iff t t t
-  | Bool Bool
-  | String Text
-  | Throw t
-  | Let Name t t
-  | t :>> t
-  | Import (NonEmpty Text)
-  | Function Name [Name] t
-  | Call t [t]
-  | Locate Span t
-  deriving (Eq, Foldable, Functor, Ord, Show, Traversable)
-
-infixl 1 :>>
+-- | Non-generic production of the recursive set of subterms.
+--
+-- This should be exactly the set of nodes which 'eval' can visit, i.e. it excludes comments, etc.
+subterms :: Term -> Set.Set Term
+subterms t = Set.insert t $ case t of
+  Module (Py.Module ss)                 -> suite ss
+  Statement (Py.Conditional cts e _)    -> foldMap (\ (c, t) -> subterms (Expr c) <> suite t) cts <> suite e
+  Statement (Py.Raise (Py.RaiseV3 e) _) -> maybe Set.empty (subterms . Expr . fst) e
+  -- FIXME: Py.RaiseV2
+  -- FIXME: whatever the second field is
+  Statement (Py.StmtExpr e _)           -> subterms (Expr  e)
+  Statement (Py.Fun _ _ _ ss _)         -> suite ss
+  -- FIXME: include 'subterms' of any default values
+  Expr (Py.Call f as _)                 -> subterms (Expr f) <> foldMap (\case { Py.ArgExpr e _ -> subterms (Expr e) ; _ -> Set.empty }) as
+  -- FIXME: support keyword args &c.
+  _                                     -> Set.empty -- TBD, and terminals
+  where
+  suite = foldMap (subterms . Statement)
```

Comment:
Note that various terminal nodes (literals, etc) are all (correctly) handled by the fallback case—but so are nonterminals we just haven't got around to yet.

I couldn't think of any generic way of implementing this which wouldn't end up being more trouble than hand-writing (and, regrettably, -maintaining) this. Oh, except I just did: an abstract interpretation of a program ends up being a kind of complicated fold. So if we were to define a domain modelling a set of terms, we could use one abstract interpretation to compute the set(s) of terms to use for dead code analysis in a second pass 😝